### DISCUSSION TITLE
Generic OAuth2 - how-to configure?

### DISCUSSION THREAD

**Discussion by chechoochoo** [requarks/wiki#5486] [Help / Questions]
👋 I am having problems with properly configure "Generic OAuth2" on my local instance;

I tried running the official release as docker container (_image tag: 2.5.280_) as well as running from sources (_current master_); tried various "Oauth2 servers" that are also running locally (i.e. https://blog.logrocket.com/implement-oauth-2-0-node-js/, https://github.com/14gasher/oauth-example) - in all cases I see "Failed to obtain access token" :(

While running wiki from sources, error is little bit more verbose:

```
InternalOAuthError: Failed to obtain access token
    at OAuth2Strategy._createOAuthError (/wiki/node_modules/passport-oauth2/lib/strategy.js:423:17)
    at /wiki/node_modules/passport-oauth2/lib/strategy.js:177:45
    at /wiki/node_modules/oauth/lib/oauth2.js:191:18
    at ClientRequest.<anonymous> (/wiki/node_modules/oauth/lib/oauth2.js:162:5)
    at ClientRequest.emit (events.js:400:28)
    at Socket.socketErrorListener (_http_client.js:475:9)
    at Socket.emit (events.js:400:28)
    at emitErrorNT (internal/streams/destroy.js:106:8)
    at emitErrorCloseNT (internal/streams/destroy.js:74:3)
    at processTicksAndRejections (internal/process/task_queues.js:82:21)
```

As there are quite a few variables that might-not-work, was hoping that someone could advise / point to potential cause 🤔 looking forward 👍 

----

note: I did try https://github.com/requarks/wiki/discussions/5149#discussioncomment-2568859 - but error is still as mentioned above

**Comment by jacodv** [comment]
@chechoochoo - I hope that you can help me?  I am trying to debug/trace what is the root cause of my errors.

I have the same error, what makes it worse is that I cannot find any logs in the running container.  I am probably doing something wrong.

UPDATE: I am also running everything locally:
* Wikijs (localhost:3000)
* Auth server (localhost:5000)
* Wikijs receives request (`http://localhost:3000/login/e48437c7-d7fa-41d0-91dc-5e7755f5232a/callback?code=CfDJ8JWYaUjmfghDq2H......`)

* I have tried to change to the config.yaml `logLevel: silly`
* I have looked for a folder `logs` or `Logs`
* Below are the errors if I swop between Generic OAuth2 and OpenId Connect / OAuth2
Oops, something went wrong...
Invalid email / username or password. 

Oops, something went wrong...
Failed to obtain access token.

**Reply by jacodv** [reply]
@chechoochoo -  I can confirm that the authorization server as well as passport-oauth2/passport works.  I have created a quick and dirty express app.  The Issue must be inside Wikijs code and not the referenced libs.  

I am not sure what is required to run / debug Wikijs.  But I am willing to help find the root cause @NGPixel 

* localhost:3000 - Express Authorization Client
* localhost:4200 - Angular UI with LoginComponent
* localhost:62294 - ASPNET Core API with OAuth2 Authorization endpoint(s)

![image](https://user-images.githubusercontent.com/5814127/193817839-596a0b55-1728-4de7-b263-b68eca639d98.png)

Here is the code for the quick and dirty:
```const express = require('express')
const session = require('express-session');
const passport = require('passport')
const OAuth2Strategy = require('passport-oauth2').Strategy;

var querystring = require('querystring');
var request = require('request');

const app = express()
app.use(session({ secret: 'SECRET' }));
const port = 3000

app.get('/', (req, res) => {
  res.send('Hello World!');
})

app.get('/authorizeStart', passport.authenticate("oauth2"));

app.get('/authorizeError', (req, res) => {
  res.send('Authorize ERROR!');
})

app.get('/authorizeSuccess', (req, res) => {
  res.send('Authorize SUCCESS :-)');
})

app.get('/authorizeCallBack',
  passport.authenticate('oauth2', { failureRedirect: '/authorizeError', session:false }),
  function(req, res) {
    // Successful authentication, redirect home.
    console.log('authorizeCallBack success');
    res.redirect('/authorizeSuccess');
  });

app.listen(port, () => {
  passport.use(new OAuth2Strategy({
    authorizationURL: 'http://localhost:4200/auth/authorize',
    tokenURL: 'http://localhost:62294/oauth',
    clientID: 'ifsdocs',
    clientSecret: 'EXAMPLE_CLIENT_SECRET',
    callbackURL: "http://localhost:3000/authorizeCallBack"
  },
  function(accessToken, refreshToken, profile, cb) {
    console.log('In passport.use', {accessToken, refreshToken, profile, cb});
    
   request({
    headers:{'Content-Type': 'application/json', 'Authorization': 'Bearer ' + accessToken},
    uri: 'http://localhost:62294/api/user/whois',
    method: 'GET'
   }, function(err,res,body){
    console.log('WhoIS:',{err,body});    
    cb(err, body);
   })

  }
  ));
  console.log(`Example app listening on port ${port}`)
})```

**Reply by chechoochoo** [reply]
@jacodv thank you for your input and I am deeply sorry for such a delay in commenting :(

TBH, we had to apply completely different approach (_=custom strategy_), as plain OAuth would not fit our use-case in the long-run

**Reply by jacodv** [reply]
Ok, @chechoochoo, do you have a suggestion for us?  We have a UAT platform where our clients play in sandboxes, and we want to use the same user base to sign the users in.  We are more than willing to just handle the sign-in and all the rest in wikijs.

**Reply by felagund** [reply]
Are you using selsigned certificates by any chance? If so, I stumbled across this: https://github.com/Requarks/wiki/discussions/3387

and indeed setting the variable helped. (I am doing it on yunohost, so more details here: https://forum.yunohost.org/t/oauth2-nextcloud-and-wiki-js-or-how-can-i-pass-a-node-variable-to-wiki-js/24058 )

**Comment by jacodv** [comment]
Worth a try.  Thanks.
